### PR TITLE
feat: introduce stream message schema and validation

### PR DESCRIPTION
**Pull request by lvrach**
# Description

From https://github.com/rudderlabs/rudder-go-kit/pull/421

A common presentation message schema https://www.notion.so/rudderstacks/Message-definition-ece5fa4c9ab4463cb90c864e17d7c36a?pvs=4, to be used by ingestion-svc, src-router, rudder-server/gateway

## Linear Ticket

[PIPE-1004](https://linear.app/rudderstack/issue/PIPE-1004/common-message-definition-schema)

resolves PIPE-1004

## Security

- [x] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

File: .github/workflows/release-please.yaml [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+name: release-please
+jobs:
+  release-please:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Extract branch name
+        shell: bash
+        run: echo "##[set-output name=branch;]$(echo ${GITHUB_REF#refs/heads/})"
+        id: extract_branch
+      - uses: google-github-actions/release-please-action@v3
```
Comment (achettyiitr) [inline .github/workflows/release-please.yaml:14]:
```suggestion
      - uses: google-github-actions/release-please-action@v4
```

File: .github/workflows/release-please.yaml [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+        with:
+          token: ${{ secrets.PAT }}
+          pull-request-title-pattern: "chore: release ${version}"
+          release-type: go
+          package-name: rudder-server
```
Comment (achettyiitr) [inline .github/workflows/release-please.yaml:19]:
```suggestion
          package-name: rudder-schemas
```

File: .github/workflows/semantic-pr.yaml [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+      - ready_for_review
+      - synchronize
+
+jobs:
+  main:
+    name: title
+    runs-on: ubuntu-latest
+    steps:
+      - 
+        uses: amannn/action-semantic-pull-request@v4
```
Comment (achettyiitr) [inline .github/workflows/semantic-pr.yaml:20]:
```suggestion
        uses: amannn/action-semantic-pull-request@v5
```

File: .github/workflows/stale-pr.yaml [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+          days-before-pr-stale: 20
+          days-before-pr-close: 7
+          stale-pr-label: 'Stale'
+
+  branches:
+      name: Cleanup old branches
+      runs-on: ubuntu-latest
+      steps:
+        - name: Checkout repository
+          uses: actions/checkout@v2
```
Comment (achettyiitr) [inline .github/workflows/stale-pr.yaml:30]:
```suggestion
          uses: actions/checkout@v4
```

File: .github/workflows/stale-pr.yaml [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+        - name: Run delete-old-branches-action
+          uses: beatlabs/delete-old-branches-action@v0.0.9
```
Comment (achettyiitr) [inline .github/workflows/stale-pr.yaml:32]:
```suggestion
          uses: beatlabs/delete-old-branches-action@v0.0.10
```

File: .github/workflows/tests.yaml [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+on:
+  push:
+    branches:
+      - main
+      - "release/*"
+  pull_request:
+jobs:
+  unit:
+    name: Unit
+    runs-on: 'ubuntu-22.04'
```
Comment (achettyiitr) [inline .github/workflows/tests.yaml:11]: Can we use latest in here?
```suggestion
    runs-on: ubuntu-latest
```

File: .github/workflows/tests.yaml [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+    steps:
+      - uses: actions/checkout@v3
+      - uses: actions/setup-go@v5
```
Comment (achettyiitr) [inline .github/workflows/tests.yaml:14]:
```suggestion
      - uses: actions/checkout@v4
      - uses: actions/setup-go@v5
```

File: go.mod [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+	github.com/gabriel-vasile/mimetype v1.4.3 // indirect
+	github.com/go-playground/locales v0.14.1 // indirect
+	github.com/go-playground/universal-translator v0.18.1 // indirect
+	github.com/leodido/go-urn v1.4.0 // indirect
+	github.com/stretchr/testify v1.9.0
+	golang.org/x/crypto v0.19.0 // indirect
+	golang.org/x/net v0.21.0 // indirect
+	golang.org/x/sys v0.17.0 // indirect
+	golang.org/x/text v0.14.0 // indirect
+)
```
Comment (achettyiitr) [inline go.mod:25]: Can we club these two?

File: go/stream/message.go [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+			MessageID:       properties[pulsarKeyMessageID],
+			RoutingKey:      properties[pulsarKeyRoutingKey],
+			WorkspaceID:     properties[pulsarKeyWorkspaceID],
+			UserID:          properties[pulsarKeyUserID],
+			SourceID:        properties[pulsarKeySourceID],
+			SourceJobRunID:  properties[pulsarKeySourceJobRunID],
+			SourceTaskRunID: properties[pulsarKeySourceTaskRunID],
+			TraceID:         properties[pulsarKeyTraceID],
+		},
+		Payload: json.RawMessage(payload),
```
Comment (achettyiitr) [inline go/stream/message.go:49]: Redundant type conversion 
```suggestion
		Payload: payload,
```

File: go/stream/message.go [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+	return map[string]string{
+		pulsarKeyMessageID:       properties.MessageID,
+		pulsarKeyRoutingKey:      properties.RoutingKey,
+		pulsarKeyWorkspaceID:     properties.WorkspaceID,
+		pulsarKeySourceID:        properties.SourceID,
+		pulsarKeyUserID:          properties.UserID,
+		pulsarKeySourceJobRunID:  properties.SourceJobRunID,
+		pulsarKeySourceTaskRunID: properties.SourceTaskRunID,
+		pulsarKeyTraceID:         properties.TraceID,
+	}, []byte(msg.Payload)
```
Comment (achettyiitr) [inline go/stream/message.go:65]: Redundant type conversion


```suggestion
	}, msg.Payload
```

File: .github/workflows/stale-pr.yaml (hunk 4 above)
Comment (achettyiitr) [reply .github/workflows/stale-pr.yaml:30]: Can we use the latest (v4) in other files as well.

**Review by achettyiitr — APPROVED** [review]

**Comment by achettyiitr** [comment]
LGTM with minor changes.

File: go/stream/message.go [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+
+type MessageProperties struct {
+	MessageID       string `json:"messageID" validate:"required"`
+	RoutingKey      string `json:"routingKey" validate:"required"`
+	WorkspaceID     string `json:"workspaceID" validate:"required"`
+	UserID          string `json:"userID" validate:"required"`
+	SourceID        string `json:"sourceID" validate:"required"`
+	SourceJobRunID  string `json:"sourceJobRunID"`  // optional
+	SourceTaskRunID string `json:"sourceTaskRunID"` // optional
+	TraceID         string `json:"traceID"`         // optional
```
Comment (fracasula) [inline go/stream/message.go:33]: Could `omitempty` be useful here?

File: go/stream/message.go [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+			SourceJobRunID:  properties[pulsarKeySourceJobRunID],
+			SourceTaskRunID: properties[pulsarKeySourceTaskRunID],
+			TraceID:         properties[pulsarKeyTraceID],
+		},
+		Payload: json.RawMessage(payload),
+	}, nil
+}
+
+// ToPulsarMessage converts a Message to a Pulsar message.
+func ToPulsarMessage(msg Message) (map[string]string, []byte) {
```
Comment (fracasula) [inline go/stream/message.go:54]: I wonder if we should concern this with the word `Pulsar` at all. I think `ToMessageProperties` would work as well and could be used for HTTP headers too. Just a thought anyway, not blocking.

File: go/stream/message.go [hunk 13]
```diff
@@ -0,0 +1,74 @@
+package stream
+
+import (
+	"encoding/json"
+
+	"github.com/go-playground/validator/v10"
```
Comment (fracasula) [inline go/stream/message.go:6]: [non-blocking] I never used a struct validator before. Is this the most popular one? Are there more available possibly with benchmarks? When I googled it I came across this (closed) [issue](https://github.com/go-validator/validator/issues/78) so I now wonder :sweat_smile:

**Review by fracasula — APPROVED** [review]

**Pull request merged**